### PR TITLE
S3655: "In" method params

### PR DESCRIPTION
**Pull request by antonioaversa**
Part 7 of task 6 of https://github.com/SonarSource/sonar-dotnet/issues/6794

Previous task: https://github.com/SonarSource/sonar-dotnet/pull/7000

Addresses:
- comments received on https://github.com/SonarSource/sonar-dotnet/pull/6996 on commits already merged into `feature/SE`, reviewed twice due to a change of base branch as explained [here](https://github.com/SonarSource/sonar-dotnet/pull/6996#discussion_r1151472642);
- comment received offline on https://github.com/SonarSource/sonar-dotnet/pull/6962#discussion_r1149128723: `The idea was to assert all (again) is to be sure that the pairing (left vs. right) doesn't go out of sync. That would ensure that it works, and that it works for a good reason. In theory, the pairing could go out of sync somewhere and it could end up as a green test by coincidence.`

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 1]
```diff
@@ -850,6 +880,69 @@ async Task WithAsyncAwait()
...
+        _ = (int)(object)(null as int?);    // Compliant
+    }
+
+    void NullableRoundTrip()
+    {
+        int? nullable = null;
+        object implicitBoxed = nullable;
+        int? unboxedNullable = (int?)implicitBoxed;
+        _ = unboxedNullable.Value;          // Noncompliant, empty
+    }
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:902]: The test case where `nullable` is a parameter is missing.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 2]
```diff
@@ -908,3 +1001,58 @@ struct StructWithValuePropertyAndCastOperators
     }
 }
 
+namespace TypeWithValueStaticProperty
+{
+    class Test
+    {
+        void Basics()
+        {
+            _ = AClassWithStaticValueProperty.Value;                                // Compliant, not on nullable value type
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:1010]: You need to learn "Null" for all the properties first. Otherwise, you would never get the rule triggered:
```suggestion
            _ = AClassWithStaticValueProperty == null ? AClassWithStaticValueProperty.Value : null;                                // Compliant, not on nullable value type
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 3]
```diff
@@ -908,3 +1001,58 @@ struct StructWithValuePropertyAndCastOperators
...
+            _ = AClassWithStaticValueProperty.Value.Value;                          // Compliant
+            _ = AClassWithStaticValueProperty.Value.Value.InstanceProperty;         // Compliant
+            _ = AClassWithStaticValueProperty.Value.Value.InstanceProperty.Value;   // Compliant
+            _ = new AClassWithInstanceValueProperty().Value;                        // Compliant
+        }
+    }
+
+    class AClassWithStaticValueProperty
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:1018]: That is quite a mouth full:
```suggestion
    class StaticValue
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 4]
```diff
@@ -908,3 +1001,58 @@ struct StructWithValuePropertyAndCastOperators
...
+    {
+        public AClassWithInstanceValueProperty InstanceProperty => new AClassWithInstanceValueProperty();
+
+        public static AClassWithInstanceValueProperty Value => new AClassWithInstanceValueProperty();
+    }
+
+    class AClassWithInstanceValueProperty
+    {
+        public AClassWithStaticValueProperty Value => new AClassWithStaticValueProperty();
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:1027]: Two changes:
* Make it an auto property because of #7007
* Add a Test with access to `this.Value` in an instance method of this class.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 5]
```diff
@@ -908,3 +1001,58 @@ struct StructWithValuePropertyAndCastOperators
...
+    }
+}
+
+namespace TypeWithValueStaticField
+{
+    class Test
+    {
+        void Basics()
+        {
+            _ = AClassWithStaticValueField.Value;                                // Compliant, not on nullable value type
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:1037]: Same here: You need to learn "Null".

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 6]
```diff
@@ -737,6 +737,39 @@ void RefParams(int? iParam)
...
+        _ = iParam.Value;       // Compliant, iParam is "in" and its value can't be modified
+    }
+
+    void InParamOfAnotherMethod(in int? iParam)
+    {
+        _ = iParam.Value;       // Compliant, unknown
+
+        MethodWithReadInParam(iParam);
+        _ = iParam.Value;       // Compliant, iParam is "in" and its value can't be modified
+    }
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:768]: We need this test case where `iParam` is a local known to be null. Then the call is done and the value is passed:
```suggestion
    void InParamOfAnotherMethod(in int? iParam)
    {
        _ = iParam.Value;       // Compliant, unknown

        MethodWithReadInParam(iParam);
        _ = iParam.Value;       // Compliant, iParam is "in" and its value can't be modified
    }
    
    void InParamToAnotherMethod()
    {
        int? arg = null;
        MethodWithReadInParam(arg);
        _ = arg.Value;       // Noncompliant, arg is "in" and its value must still be null here
    }    
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 1 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: Defined [here](https://github.com/SonarSource/sonar-dotnet/pull/7006/files#diff-9e6c038bf5ad1e53d1b547b8870231cc2887c9c53296f73e3eb811f51613bdedR912)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 2 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: [DRAFT]

A bit of context: these tests have been written in the context of https://github.com/SonarSource/sonar-dotnet/pull/7000, to solve the `NullReferenceException` instances encountered on peach. The issue was due to custom `Value` properties (user-defined, not `Nullable<T>.Value`) generating `PropertyReferenceOperationWrapper` instances having `Instance` equal to `null`, and that is because they are `static`.

This PR is still in Draft state, assigned to me on the "In Progress" column of the Kanban board: not yet ready to be reviewed because not yet rebased onto the latest `feature/SE`. As a result of that, it now contains commits that have been reviewed and merged into `feature/SE`. So a good chunk of reviewed code has already been reviewed in https://github.com/SonarSource/sonar-dotnet/pull/7000 and merged into the `feature/SE` branch.

Now, coming to the specific problem: since these tests were written to reproduce and fix NRE on peach, they don't aim at reproducing non-compliance. By the way, that would be hard to get for chained `.Value` dereferences anyway, since we don't learn on those properties (since they are not auto-properties).

Scenarios about learning on mutable fields have already been tested in the `MutableField` class, and I can add a dedicated class on `MutableProperties`, to test auto-properties, if you feel that is necessary. My understanding is that, however, those tests would have the correct behavior only after https://github.com/SonarSource/sonar-dotnet/pull/7007 is merged into `feature/SE`, because we don't currently learn from auto-properties.

So, for this PR specifically, I have added

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 4 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: I think [this](https://github.com/SonarSource/sonar-dotnet/pull/7006#discussion_r1154638648) comment applies here too.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 5 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: I think [this](https://github.com/SonarSource/sonar-dotnet/pull/7006#discussion_r1154638648) comment applies here too.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 6 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:768]: Added, I called it `LocalAsInParamToAnotherMethod`, to make clear that the `in` param is `in` for the other method, and not for the method which is being executed.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 7]
```diff
@@ -737,6 +737,46 @@ void RefParams(int? iParam)
     }
 }
 
+class InParams
+{
+    void InParamOfALocalFunction(int? iParam)
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:742]: This test and the `LocalAsInParamToAnotherMethod` test are almost identical. I would remove the local function here and use `MethodWithReadInParam` instead. LocalAsInParamToAnotherMethod can be removed then. `MethodWithReadInParam` can be named InParam or something like that. I don't see how "Method" "With" and "Read" help here. You should instead add the optional `in` modifier on the call side: `InParam(in iParam)`. That makes it obvious that the test is about the modifier's semantics.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
We can clean-up the `in` test cases a bit.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 7 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:742]: Add the optional `in` modifier on the call side: done.

Remove the local function and use `MethodWithReadInParam`: done. The idea here was to test both local functions and class instance methods, and not make any assumption on whether the underlying symbolic execution engine treats the two structures differently. I guess there would or will be differences in how inter-procedural execution is done, across methods, non-static local functions and static local functions. 

`MethodWithReadInParam` rename into `InParam`: done.

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM
